File: demos/linkedlistdemo/linked_list_demo/src/main/java/com/linkedlist/app/LinkedListIterator.java
package com.linkedlist.app;

import java.util.Iterator;

/**
 * Iterator implementation for the LinkedList
 */
public class LinkedListIterator implements Iterator<Node> {
    private Node current;

    public LinkedListIterator(Node start) {
        this.current = start;
    }

    @Override
    public boolean hasNext() {
        return current != null;
    }

    @Override
    public Node next() {
        Node temp = current;
        current = current.next;
        return temp;
    }
}

File: demos/linkedlistdemo/linked_list_demo_kotlin_gradle/src/main/kotlin/com/linkedlist/app/LinkedListIterator.kt
package com.linkedlist.app

/**
 * Iterator implementation for the LinkedList
 */
class LinkedListIterator(private var current: Node?) : Iterator<Node> {
    
    override fun hasNext(): Boolean {
        return current != null
    }

    override fun next(): Node {
        val temp = current ?: throw NoSuchElementException()
        current = current?.next
        return temp
    }
}
